### PR TITLE
Implement custom tags provider for HTTP metrics

### PR DESCRIPTION
**Pull request by jotak**
- Allows to implement a CustomTagsProvider interface to provide tags out
of HttpRequest objects.
- Add tests

Fixes https://github.com/vert-x3/vertx-micrometer-metrics/issues/89

### CODE REVIEW

**Comment by jotak** [comment]
Hey @vietj  perhaps it needs a little bit of discussion prior to merging.
To summarize, the feature is about having the ability to customize metric tags out of request data, in particular headers. Basically, tags can be injected for every request based on some request or headers mapping function.

Here I've implemented the most straightforward way, which is by having the user providing a `CustomTagsProvider` mapper into config, something like that:

```java
    vertx = Vertx.vertx(new VertxOptions().setMetricsOptions(new MicrometerMetricsOptions()
      .setRequestsTagsProvider(req -> {
        String user = req.headers().get("user");
        return Collections.singletonList(Tag.of("user", user));
      })
      .setPrometheusOptions(new VertxPrometheusOptions().setEnabled(true))
      .setEnabled(true)));
```

* Pros: it's flexible; The user can create tags the way she wants out of an `HttpRequest` object
* Cons: non-serializable, the `setRequestsTagsProvider` option is executable code, cannot be set up via json

So, we can imagine an alternative where a similar feature would be achieved by providing a (serializable) option, e.g. a list of header keys or a list of pairs [header key, tag name] that would be used for a mapping of headers to tags. But it would be less flexible, no possibility of combining headers, or using other info from `HttpRequest`, or having more complex behaviours. So people may feel limited.

WDYT? Are we good with my first idea; and eventually we can iterate later to enhance the feature by adding serializable options if that's what users want?

**Comment by jotak** [comment]
PS: CI failure is related to https://github.com/vert-x3/vertx-micrometer-metrics/pull/123

**Comment by narzach** [comment]
@jotak This is great, thank you for taking the time to make this happen!

IMO your approach is exactly what we need. The serializable option would never be as useful as this is, and IMO would not provide enough value to be worth doing. This gives us full control to set tags based on the context request. For my use cases, I need the ability to do exactly what you're doing, extracting values from HTTP path parameters, query string, or headers, and then injecting corresponding tags. Static tags that are added per app will be useful, yes, but that is trivial and can be done with your setRequestsTagsProvider() approach.

Curious if anyone else has had a chance to look at this?

File: src/main/java/io/vertx/micrometer/impl/CustomTagsProvider.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+ * limitations under the License.
+ */
+package io.vertx.micrometer.impl;
+
+import io.micrometer.core.instrument.Tag;
+
+/**
+ * @author Joel Takvorian
+ */
+public interface CustomTagsProvider<T> {
```
Comment (vietj) [inline src/main/java/io/vertx/micrometer/impl/CustomTagsProvider.java:24]: this interface is in an impl package, I think instead we should rely on Java function types to do this, i.e the provider should simply be a `Function<T, Iterable<Tag>>`

File: src/main/java/io/vertx/micrometer/MicrometerMetricsOptions.java [hunk 2]
```diff
@@ -66,6 +68,7 @@
   private VertxJmxMetricsOptions jmxMetricsOptions;
   private boolean jvmMetricsEnabled;
   private MetricsNaming metricsNaming;
+  private CustomTagsProvider<HttpRequest> requestsTagsProvider;
```
Comment (vietj) [inline src/main/java/io/vertx/micrometer/MicrometerMetricsOptions.java:71]: instead use `Function<HttpRequest, Iterable<Tag>>`

**Review by vietj — CHANGES_REQUESTED** [review]
use a function instead of a specific new type

**Comment by jotak** [comment]
@vietj done

**Review by vietj — CHANGES_REQUESTED** [review]
Documentation

File: src/main/asciidoc/index.adoc [hunk 3]
```diff
@@ -369,6 +369,19 @@ Here is an example to replace the actual `path` label of HTTP requests with a ge
 
 NOTE: Matchers use MeterFilters under the hood.
 
+=== Custom tags provider
+
+It is possible to define a function that generates additional tags (or labels) for HTTP server metrics.
```
Comment (vietj) [inline src/main/asciidoc/index.adoc:374]: _You can define_

File: src/main/asciidoc/index.adoc [hunk 4]
```diff
@@ -369,6 +369,19 @@ Here is an example to replace the actual `path` label of HTTP requests with a ge
...
+This function takes an {@link io.vertx.core.spi.observability.HttpRequest} object as a parameter, and returns
```
Comment (vietj) [inline src/main/asciidoc/index.adoc:375]: Such function

Comment (jotak) [reply src/main/asciidoc/index.adoc]: done

File: src/main/asciidoc/index.adoc (hunk 3 above)
Comment (jotak) [reply src/main/asciidoc/index.adoc]: done

**Pull request merged**